src/main/scala/chisel3/util/circt/DPI.scala

```diff
@@ -0,0 +1,141 @@
+    * their values will not be modified in that clock.
+    *
+    * Please refer https://github.com/llvm/circt/blob/main/docs/Dialects/FIRRTL/FIRRTLIntrinsics.md#dpi-intrinsic-abi for DPI function ABI.
+    * @example {{{
+    * val a = RawClockedNonVoidFunctionCall("dpi_func_foo", UInt(1.W), clock, enable, b, c)
+    * }}}
+    */
+  def apply[T <: Data](
+    functionName: String,
+    ret:          => T,
+    inputNames:   Option[Seq[String]] = None,
+    outputName:   Option[String] = None
+  )(clock:        Clock,
+    enable:       Bool,
+    data:         Data*
+  ): T = {
+    IntrinsicExpr(
+      "circt_dpi_call",
+      ret,
+      GetDPIParams(functionName, true, inputNames, outputName): _*
+    )(
+      (Seq(clock, enable) ++ data): _*
+    )
+  }
+}
+
+object RawUnclockedNonVoidFunctionCall {
+
+  /** Creates an intrinsic that calls non-void DPI function for its input value changes.
+    * The DPI call is considered as a combinational logic.
+    *
+    * `enable` operand is used to conditionally call the DPI since DPI call could be quite
+    * more expensive than native constructs.
+    * When `enable` is false, results of unclocked calls are undefined and evaluated into X.
+    *
+    * Please refer https://github.com/llvm/circt/blob/main/docs/Dialects/FIRRTL/FIRRTLIntrinsics.md#dpi-intrinsic-abi for DPI function ABI.
+    * @example {{{
+    * val a = RawUnclockedNonVoidFunctionCall("dpi_func_foo", UInt(1.W), enable, b, c)
+    * }}}
+    */
+  def apply[T <: Data](
+    functionName: String,
+    ret:          => T,
+    inputNames:   Option[Seq[String]] = None,
+    outputName:   Option[String] = None
+  )(enable:       Bool,
+    data:         Data*
+  ): T = {
+    IntrinsicExpr(
+      "circt_dpi_call",
+      ret,
+      GetDPIParams(functionName, false, inputNames, outputName): _*
+    )(
+      (Seq(enable) ++ data): _*
+    )
+  }
+}
+
+object RawClockedVoidFunctionCall {
+
+  /** Creates an intrinsic that calls void DPI function at its clock posedge.
+    *
+    * Please refer https://github.com/llvm/circt/blob/main/docs/Dialects/FIRRTL/FIRRTLIntrinsics.md#dpi-intrinsic-abi for DPI function ABI.
+    * @example {{{
+    * RawClockedVoidFunctionCall("dpi_func_foo", UInt(1.W), clock, enable, b, c)
+    * }}}
+    */
+  def apply(
+    functionName: String,
+    inputNames:   Option[Seq[String]] = None
+  )(clock:        Clock,
+    enable:       Bool,
+    data:         Data*
+  ): Unit = {
+    Intrinsic("circt_dpi_call", GetDPIParams(functionName, true, inputNames): _*)(
+      (Seq(clock, enable) ++ data): _*
+    )
+  }
+}
+
+// A common trait for DPI functions.
+trait DPIFunctionImport {
+  val functionName: String
+  val inputNames: Option[Seq[String]] = None
+}
+
+// Base trait for a non-void function that returns `T`.
+trait DPINonVoidFunctionImport[T <: Data] extends DPIFunctionImport {
+  val ret:     T
+  val clocked: Boolean
+  val outputName: Option[String] = None
```

Comment:
```suggestion
trait DPIFunctionImport {
  def functionName: String
  def inputNames: Option[Seq[String]] = None
}

// Base trait for a non-void function that returns `T`.
trait DPINonVoidFunctionImport[T <: Data] extends DPIFunctionImport {
  def ret:     T
  def clocked: Boolean
  def outputName: Option[String] = None
```

Virtual methods should be defs.